Ошибка сервиса при запросе песен с валидным токеном

Приветствую! Ровно месяц назад нашёл Вашу библиотеку и адаптировал под неё код своего бота, получение и поиск аудиозаписей работали исправно. На следующий день аккаунт ВК моего бота заблокировали из-за "подозрительной активности" и заняться этим вопросом я смог только сейчас. Я провёл необходимые мероприятия и обнаружил, что бот больше не хочет получать аудиозаписи пользователя и искать музыку, в логах была ошибка. Я воспроизвёл код примера, попробовал сначала свой токен (с моим `user_agent`), затем сгенерировал дважды через вспомогательный класс и результат один - ошибка. Вот код, он взят из примера, запущен в интерпретаторе Python:

```
login = "38012345678"
password = "vk_password"
tokenReceiver = TokenReceiver(login, password)
tokenReceiver.auth()
2025-01-02 00:07:12,127 | [vkpymusic.TokenReceiver | (TokenReceiver.py) .auth(193)] [INFO] Token was received!
True
tokenReceiver.get_token()
2025-01-02 00:07:29,321 | [vkpymusic.TokenReceiver | (TokenReceiver.py) .get_token(210)] [INFO] vk1.a......xyz
'vk1.a......xyz'
tokenReceiver.save_to_config()
File already exist! Enter "OK" for rewriting it
OK
2025-01-02 00:07:40,242 | [vkpymusic.TokenReceiver | (TokenReceiver.py) .save_to_config(234)] [INFO] Token was saved!
service = Service.parse_config()
tracks = service.search_songs_by_text('idfc tiktok remix')
2025-01-02 00:07:51,918 | [vkpymusic.Service | (Service.py) .search_songs_by_text(338)] [INFO] Request by text: "idfc tiktok remix" в количестве 3
2025-01-02 00:07:52,182 | [vkpymusic.Service | (Service.py) .__response_to_songs(167)] [ERROR] 'response'
2025-01-02 00:07:52,182 | [vkpymusic.Service | (Service.py) .search_songs_by_text(344)] [ERROR] cannot access local variable 'items' where it is not associated with a value
```

Поздравляю! Вижу, что было сделано много работы!

> Версию обновил, питон 3.8 поддерживает

Ну так и раньше поддерживал, я ранее поднимал вопрос насчёт поддержки 3.12, а на Manjaro Linux сейчас в репозиториях и вовсе 3.13. 

Бегло ознакомился, вроде работает. Заметил интересную странность - почему-то `save_music` не хочет сохранять, хотя некоторые из нижеприведённых ссылок точно проигрываются в браузере:
```
>>> x = service.search_songs_by_text("Сплин что ты будешь делать?", 10)
2025-02-18 19:42:00,110 | service.py(559) | service.search_songs_by_text(...) | [INFO] Request by text: "Сплин что ты будешь делать?" в количестве 10
2025-02-18 19:42:00,521 | service.py(569) | service.search_songs_by_text(...) | [INFO] Found 10 songs
>>> for i in x:
...     print(i.artist, i.title, not i.url)
... 
Андрей Мариничев и Алексей Боголюбов Что ты будешь делать (2006), сл. и муз. А. Васильев (Сплин) False
Сплин Что ты будешь делать? False
Сплин Война False
Сплин Что ты будешь делать? (1996, Коллекционер оружия) False
Сплин Что ты будешь делать? False
Сплин Что ты будешь делать (аккустика) False
AI-Project Сплин - Что ты будешь делать False
Сплин Слин_Что ты будешь делать False
Сплин Что ты будешь делать? False
#ПОСАДОЧНИКИ сплин - что будешь делать ты? False
>>> service.save_music(x[0])
2025-02-18 19:42:35,736 | service.py(920) | service.save_music(...) | [WARNING] File with name Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов.mp3 exists. Overwrite it? (Y/n)
Y
2025-02-18 19:42:38,744 | service.py(927) | service.save_music(...) | [INFO] Downloading Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов...
2025-02-18 19:42:38,806 | service.py(933) | service.save_music(...) | [INFO] Success! Music was downloaded in '/home/amnesia/Music/Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов.mp3'
'/home/amnesia/Music/Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов.mp3'
>>> service.save_music(x[1])
2025-02-18 19:42:46,794 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать - Сплин: timed out
>>> service.save_music(x[2])
2025-02-18 19:43:01,986 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Война - Сплин: timed out
>>> service.save_music(x[3])
2025-02-18 19:43:09,242 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать 1996 Коллекционер оружия - Сплин: timed out
>>> service.save_music(x[4])
2025-02-18 19:44:12,206 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать - Сплин: timed out
>>> service.save_music(x[5])
2025-02-18 19:44:14,837 | service.py(907) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать аккустика - Сплин: 302
>>> service.save_music(x[6])
2025-02-18 19:44:25,743 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Сплин  Что ты будешь делать - AIProject: timed out
>>> service.save_music(x[7])
2025-02-18 19:44:32,905 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Слин_Что ты будешь делать - Сплин: timed out
>>> service.save_music(x[8])
2025-02-18 19:44:40,819 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать - Сплин: timed out
>>> service.save_music(x[9])
2025-02-18 19:44:49,200 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading сплин  что будешь делать ты - ПОСАДОЧНИКИ: timed out
>>> 
```

> Поздравляю! Вижу, что было сделано много работы!

🎉🎉🎉

> > Версию обновил, питон 3.8 поддерживает
> 
> Ну так и раньше поддерживал, я ранее поднимал вопрос насчёт поддержки 3.12, а на Manjaro Linux сейчас в репозиториях и вовсе 3.13. 

Посмотрю по поводу 3.13

> Бегло ознакомился, вроде работает. Заметил интересную странность - почему-то `save_music` не хочет сохранять, хотя некоторые из нижеприведённых ссылок точно проигрываются в браузере:
> ```
> >>> x = service.search_songs_by_text("Сплин что ты будешь делать?", 10)
> 2025-02-18 19:42:00,110 | service.py(559) | service.search_songs_by_text(...) | [INFO] Request by text: "Сплин что ты будешь делать?" в количестве 10
> 2025-02-18 19:42:00,521 | service.py(569) | service.search_songs_by_text(...) | [INFO] Found 10 songs
> >>> for i in x:
> ...     print(i.artist, i.title, not i.url)
> ... 
> Андрей Мариничев и Алексей Боголюбов Что ты будешь делать (2006), сл. и муз. А. Васильев (Сплин) False
> Сплин Что ты будешь делать? False
> Сплин Война False
> Сплин Что ты будешь делать? (1996, Коллекционер оружия) False
> Сплин Что ты будешь делать? False
> Сплин Что ты будешь делать (аккустика) False
> AI-Project Сплин - Что ты будешь делать False
> Сплин Слин_Что ты будешь делать False
> Сплин Что ты будешь делать? False
> #ПОСАДОЧНИКИ сплин - что будешь делать ты? False
> >>> service.save_music(x[0])
> 2025-02-18 19:42:38,744 | service.py(927) | service.save_music(...) | [INFO] Downloading Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов...
> 2025-02-18 19:42:38,806 | service.py(933) | service.save_music(...) | [INFO] Success! Music was downloaded in '/home/user/Music/Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов.mp3'
> '/home/user/Music/Что ты будешь делать 2006 сл и муз А Васильев Сплин - Андрей Мариничев и Алексей Боголюбов.mp3'
> >>> service.save_music(x[1])
> 2025-02-18 19:42:46,794 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать - Сплин: timed out
> >>> service.save_music(x[2])
> 2025-02-18 19:43:01,986 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Война - Сплин: timed out
> >>> service.save_music(x[3])
> 2025-02-18 19:43:09,242 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать 1996 Коллекционер оружия - Сплин: timed out
> >>> service.save_music(x[4])
> 2025-02-18 19:44:12,206 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать - Сплин: timed out
> >>> service.save_music(x[5])
> 2025-02-18 19:44:14,837 | service.py(907) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать аккустика - Сплин: 302
> >>> service.save_music(x[6])
> 2025-02-18 19:44:25,743 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Сплин  Что ты будешь делать - AIProject: timed out
> >>> service.save_music(x[7])
> 2025-02-18 19:44:32,905 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Слин_Что ты будешь делать - Сплин: timed out
> >>> service.save_music(x[8])
> 2025-02-18 19:44:40,819 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading Что ты будешь делать - Сплин: timed out
> >>> service.save_music(x[9])
> 2025-02-18 19:44:49,200 | service.py(910) | service.save_music(...) | [ERROR] Error while downloading сплин  что будешь делать ты - ПОСАДОЧНИКИ: timed out
> >>> 
> ```

Кстати, интересно. Первый трек сохранился успешно, потом 4 ушло в тайм-аут, шестой - вообще 302 выдал. А оставшиеся тоже в тайм-ауте.

Такой вопрос: не менялся ли каким-либо образом айпи во время сохранения?

Знаю, что ссылки привязываются к айпишнику токена и/или запросу получения треков и действуют какое-то время (15 мин, если не ошибаюсь)

--- a/vkpymusic/service.py
+++ b/vkpymusic/service.py
@@ -229,7 +229,7 @@ class Service:
             user_id (str | int): VK user id. (NOT USERNAME! vk.com/id*******).
 
         Returns:
-            int: count of all user's songs.
+            int: count of all user's songs or -1 if access denied.
 
         Raises:
             VkApiException: If the response contains an error.
@@ -241,7 +241,24 @@ class Service:
         response: VkApiResponse = make_request(request)
         songs_count: int = response.data
         self.logger.info(f"Count of user's songs: {songs_count}")
-        return songs_count
+
+        if songs_count != 0:
+            return songs_count
+
+        # If count of songs is 0, it can be due to access
+        # denied to user's songs. So check this case.
+        self.logger.info(f"Trying to get songs by user id: {user_id}")
+        try:
+            self.get_songs_by_userid(user_id, 1)
+            return songs_count
+        except VkApiException as e:
+            # If error code is 201, it means access denied to user's songs.
+            if e.error_code == 201:
+                self.logger.warning(
+                    f"Access denied to user's songs."
+                )
+                return -1
+            raise
 
     async def get_count_by_user_id_async(self, user_id: Union[str, int]) -> int:
         """
@@ -251,7 +268,7 @@ class Service:
             user_id (str | int): VK user id. (NOT USERNAME! vk.com/id*******).
 
         Returns:
-            int: count of all user's songs.
+            int: count of all user's songs or -1 if access denied.
 
         Raises:
             VkApiException: If the response contains an error.
@@ -263,7 +280,24 @@ class Service:
         response: VkApiResponse = await make_request_async(request)
         songs_count: int = response.data
         self.logger.info(f"Count of user's songs: {songs_count}")
-        return songs_count
+
+        if songs_count != 0:
+            return songs_count
+
+        # If count of songs is 0, it can be due to access
+        # denied to user's songs. So check this case.
+        self.logger.info(f"Trying to get songs by user id: {user_id}")
+        try:
+            await self.get_songs_by_userid_async(user_id, 1)
+            return songs_count
+        except VkApiException as e:
+            # If error code is 201, it means access denied to user's songs.
+            if e.error_code == 201:
+                self.logger.warning(
+                    f"Access denied to user's songs."
+                )
+                return -1
+            raise
 
     # Songs section
     def get_songs_by_userid(

--- a/vkpymusic/service.py
+++ b/vkpymusic/service.py
@@ -5,12 +5,20 @@ This module contains the main class 'Service' for working with VK API.
 import os
 import configparser
 import logging
-from typing import Dict, Optional, Union, List
+from typing import Optional, Union, List
 
 import aiofiles
 from httpx import Client, AsyncClient, Response
 
-from .vk_api import VkApiRequestBuilder, VkApiRequest, VkApiResponse, VkApiException
+from .vk_api import (
+    VkApiRequestBuilder, 
+    VkApiRequest, 
+    VkApiResponse, 
+    VkApiException,
+
+    make_request,
+    make_request_async,
+)
 from .models import Song, Playlist, UserInfo
 from .utils import Converter, create_logger
 
@@ -109,89 +117,6 @@ class Service:
         request.fill_token(self.__token)
         request.fill_user_agent(self.user_agent)
 
-    @classmethod
-    def validate_and_parse_response(cls, api_response: Response) -> VkApiResponse:
-        """
-        Validate and parse response from VK API.
-
-        Args:
-            response (VkApiResponse): Response from VK API.
-        """
-        # If we have an error while making a request
-        if api_response.status_code != 200:
-            cls.logger.error(f"Error {api_response.status_code}: {api_response.text}")
-            raise VkApiException(
-                f"Error {api_response.status_code}: {api_response.text}"
-            )
-
-        # Convert response to json
-        api_response_data: Dict = api_response.json()
-
-        # Use None and validate by equals to None,
-        # because we can have an empty payload or 0 or False
-        payload: Dict = api_response_data.get("response", None)
-        error: Dict = api_response_data.get("error", None)
-
-        # If we have payload, return it
-        if payload is not None:
-            return VkApiResponse(payload)
-        # If we have an error, raise an exception
-        elif error is not None:
-            raise VkApiException(
-                error.get("error_code", 0),
-                error.get("error_msg", "Unknown error"),
-                error,
-            )
-        # If we have no payload and no error, raise an exception
-        else:
-            raise VkApiException(0, "Unknown error", {})
-
-    @classmethod
-    def make_request(cls, request: VkApiRequest) -> VkApiResponse:
-        """
-        Makes a synchronous request to the VK API.
-
-        Args:
-            request (VkApiRequest): The request to make.
-
-        Returns:
-            VkApiResponse: The response from the VK API.
-
-        Raises:
-            VkApiException: If the response status code is not 200.
-        """
-        with Client() as client:
-            response: Response = client.get(
-                url=request.url, headers=request.headers, params=request.params
-            )
-
-        vk_response: VkApiResponse = cls.validate_and_parse_response(response)
-
-        return vk_response
-
-    @classmethod
-    async def make_request_async(cls, request: VkApiRequest) -> VkApiResponse:
-        """
-        Makes an asynchronous request to the VK API.
-
-        Args:
-            request (VkApiRequest): The request to make.
-
-        Returns:
-            VkApiResponse: The response from the VK API.
-
-        Raises:
-            VkApiException: If the response status code is not 200.
-        """
-        async with AsyncClient() as client:
-            response: Response = await client.get(
-                url=request.url, headers=request.headers, params=request.params
-            )
-
-        vk_response: VkApiResponse = cls.validate_and_parse_response(response)
-
-        return vk_response
-
     ##############################################
     # METHODS FOR WORKING WITH TOKEN AND USER INFO
     def get_user_info(self) -> UserInfo:
@@ -207,7 +132,7 @@ class Service:
         self.logger.info("Getting user info...")
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_profile_info()
         request.fill_token(self.__token)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         user_info: UserInfo = Converter.response_to_userinfo(response)
         self.logger.info(f"User info: {user_info}")
         return user_info
@@ -225,7 +150,7 @@ class Service:
         self.logger.info("Getting user info...")
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_profile_info()
         request.fill_token(self.__token)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         user_info: UserInfo = Converter.response_to_userinfo(response)
         self.logger.info(f"User info: {user_info}")
         return user_info
@@ -245,7 +170,7 @@ class Service:
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_profile_info()
         request.fill_token(token)
         try:
-            cls.make_request(request)
+            make_request(request)
             cls.logger.info("Token is valid!")
             return True
         except VkApiException as e:
@@ -267,7 +192,7 @@ class Service:
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_profile_info()
         request.fill_token(token)
         try:
-            await cls.make_request_async(request)
+            await make_request_async(request)
             cls.logger.info("Token is valid!")
             return True
         except VkApiException as e:
@@ -313,7 +238,7 @@ class Service:
         self.logger.info(f"Request by user: {user_id}")
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_count(user_id)
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs_count: int = response.data
         self.logger.info(f"Count of user's songs: {songs_count}")
         return songs_count
@@ -335,7 +260,7 @@ class Service:
         self.logger.info(f"Request by user: {user_id}")
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_count(user_id)
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs_count: int = response.data
         self.logger.info(f"Count of user's songs: {songs_count}")
         return songs_count
@@ -364,7 +289,7 @@ class Service:
             user_id, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -395,7 +320,7 @@ class Service:
             user_id, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -433,7 +358,7 @@ class Service:
             user_id, count, offset, playlist_id, access_key
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -471,7 +396,7 @@ class Service:
             user_id, count, offset, playlist_id, access_key
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -501,7 +426,7 @@ class Service:
             playlist.owner_id, count, offset, playlist.playlist_id, playlist.access_key
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -531,7 +456,7 @@ class Service:
             playlist.owner_id, count, offset, playlist.playlist_id, playlist.access_key
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -561,7 +486,7 @@ class Service:
             text, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -591,7 +516,7 @@ class Service:
             text, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs: List[Song] = Converter.response_to_songs(response)
         if len(songs) == 0:
             self.logger.info("No results found ._.")
@@ -623,7 +548,7 @@ class Service:
             user_id, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         playlists: List[Playlist] = Converter.response_to_playlists(response)
         return playlists
     
@@ -650,7 +575,7 @@ class Service:
             user_id, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         playlists: List[Playlist] = Converter.response_to_playlists(response)
         return playlists
 
@@ -677,7 +602,7 @@ class Service:
             text, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         playlists: List[Playlist] = Converter.response_to_playlists(response)
         return playlists
 
@@ -704,7 +629,7 @@ class Service:
             text, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         playlists: List[Playlist] = Converter.response_to_playlists(response)
         return playlists
 
@@ -732,7 +657,7 @@ class Service:
             text, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         playlists: List[Playlist] = Converter.response_to_playlists(response)
         return playlists
     
@@ -760,7 +685,7 @@ class Service:
             text, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         playlists: List[Playlist] = Converter.response_to_playlists(response)
         return playlists
 
@@ -782,7 +707,7 @@ class Service:
         self.logger.info("Request popular songs")
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_popular(count, offset)
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs: List[Song] = Converter.response_to_popular(response)
         return songs
 
@@ -803,7 +728,7 @@ class Service:
         self.logger.info("Request popular songs")
         request: VkApiRequest = VkApiRequestBuilder.build_req_get_popular(count, offset)
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs: List[Song] = Converter.response_to_popular(response)
         return songs
 
@@ -837,7 +762,7 @@ class Service:
             user_id, song_id, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = self.make_request(request)
+        response: VkApiResponse = make_request(request)
         songs: List[Song] = Converter.response_to_songs(response)
         return songs
     
@@ -870,7 +795,7 @@ class Service:
             user_id, song_id, count, offset
         )
         self.fill_user_agent_and_token(request)
-        response: VkApiResponse = await self.make_request_async(request)
+        response: VkApiResponse = await make_request_async(request)
         songs: List[Song] = Converter.response_to_songs(response)
         return songs
 

--- a/vkpymusic/utils/__init__.py
+++ b/vkpymusic/utils/__init__.py
@@ -3,7 +3,7 @@ This module contains utilities for conversion and logging.
 
 Classes:
     Converter: A class for performing various conversion operations.
-    get_logger: A function for getting or creating a logger.
+    create_logger: A function for getting or creating a logger.
 """
 
 from .converter import Converter

--- a/vkpymusic/utils/converter.py
+++ b/vkpymusic/utils/converter.py
@@ -2,8 +2,7 @@
 This module contains the Converter class.
 """
 
-import json
-from typing import List, Optional
+from typing import List
 
 from ..vk_api import VkApiResponse
 from ..models import Song, Playlist, UserInfo

--- a/vkpymusic/vk_api/__init__.py
+++ b/vkpymusic/vk_api/__init__.py
@@ -7,6 +7,10 @@ Classes:
     VkApiResponse: A class that represents a response.
     VkApiException: A class that represents an error.
 
+Functions:
+    make_request: A function that makes a synchronous request to VK API.
+    make_request_async: A function that makes an asynchronous request to VK API.
+
 Constants:
     BASE_URL: A constant that stores the base URL for VK API.
     DEFAULT_PARAMS: A constant that stores the default parameters for VK API.
@@ -16,6 +20,10 @@ from .vk_api_request import VkApiRequest
 from .vk_api_request_builder import VkApiRequestBuilder
 from .vk_api_response import VkApiResponse
 from .vk_api_exception import VkApiException
+from .vk_api_extensions import (
+    make_request,
+    make_request_async,
+)
 from .vk_api_params import (
     BASE_URL,
     DEFAULT_PARAMS,
@@ -26,6 +34,9 @@ __all__ = [
     "VkApiRequestBuilder",
     "VkApiResponse",
     "VkApiException",
+
+    "make_request",
+    "make_request_async",
 
     "BASE_URL",
     "DEFAULT_PARAMS",

--- a/vkpymusic/vk_api/vk_api_extensions.py
+++ b/vkpymusic/vk_api/vk_api_extensions.py
@@ -1,0 +1,108 @@
+"""
+This module contains functions for prepare
+and making requests to the VK API.
+"""
+
+from typing import Dict
+
+from httpx import Client, AsyncClient, Response
+
+from ..vk_api import VkApiRequest, VkApiResponse, VkApiException
+
+
+def make_request(request: VkApiRequest) -> VkApiResponse:
+    """
+    Makes a synchronous request to the VK API.
+
+    Args:
+        request (VkApiRequest): The request to make.
+
+    Returns:
+        VkApiResponse: The response from the VK API.
+
+    Raises:
+        VkApiException: If the response status code is not 200.
+    """
+    with Client() as client:
+        response: Response = client.get(
+            url=request.url, headers=request.headers, params=request.params
+        )
+
+    vk_response: VkApiResponse = _validate_and_parse_response(response)
+
+    return vk_response
+
+
+async def make_request_async(request: VkApiRequest) -> VkApiResponse:
+    """
+    Makes an asynchronous request to the VK API.
+
+    Args:
+        request (VkApiRequest): The request to make.
+
+    Returns:
+        VkApiResponse: The response from the VK API.
+
+    Raises:
+        VkApiException: If the response status code is not 200.
+    """
+    async with AsyncClient() as client:
+        response: Response = await client.get(
+            url=request.url, headers=request.headers, params=request.params
+        )
+
+    vk_response: VkApiResponse = _validate_and_parse_response(response)
+
+    return vk_response
+
+
+def _validate_and_parse_response(
+    api_response: Response
+) -> VkApiResponse:
+    """
+    Validate and parse response from VK API.
+
+    Args:
+        response (VkApiResponse): Response from VK API.
+        logger (logging.Logger): Logger instance.
+    
+    Returns:
+        VkApiResponse: Parsed response from VK API.
+    """
+    # Convert response to json
+    api_response_data: Dict = api_response.json()
+
+    # Use None and validate by equals to None,
+    # because we can have an empty payload or 0 or False
+    payload: Dict = api_response_data.get("response", None)
+    error: Dict = api_response_data.get("error", None)
+
+    # If we have payload, return it
+    if payload is not None:
+        return VkApiResponse(payload)
+    # If we have an error, raise an exception
+    elif error is not None:
+        # If it's a string (OMG, VK API, why?), save entire request
+        if isinstance(error, str):
+            raise VkApiException(
+                0,
+                error,
+                api_response_data,
+            )
+        # If it's a dict, raise an exception with error code and message
+        elif isinstance(error, dict):
+            raise VkApiException(
+                error.get("error_code", 0),
+                error.get("error_msg", "Unknown error"),
+                error,
+            )
+        # If it's something else, raise an exception
+        else:
+            raise VkApiException(
+                0,
+                "Unknown error",
+                api_response_data,
+            )
+    
+    # else, return the response (thank you, VK API for different payloads)
+    return VkApiResponse(api_response_data)

--- a/vkpymusic/vk_api/vk_api_request_builder.py
+++ b/vkpymusic/vk_api/vk_api_request_builder.py
@@ -1,23 +1,99 @@
 """
-This module contains the 'VkApiRequestBuilder' class 
+This module contains the 'VkApiRequestBuilder' class
 for building requests to VK API.
 """
 
-from typing import Dict, Optional
+from typing import Dict, Tuple, Optional, Union
+
+from ..client import Client
 from .vk_api_request import VkApiRequest
-from .vk_api_params import BASE_URL, DEFAULT_PARAMS
+from .vk_api_params import BASE_URL, DEFAULT_PARAMS, DEFAULT_VERSION
 
 
 class VkApiRequestBuilder:
     """
     A class for building requests to VK API.
     """
+
+    # Auth section
+
     @staticmethod
-    def build_from_base_request(
-        method: str, 
-        url: str, 
-        params: dict
+    def build_req_auth(
+        login: str,
+        password: str,
+        client: Client,
+        code: Optional[str] = None,
+        captcha: Optional[Tuple[str, str]] = None,
     ) -> VkApiRequest:
+        """
+        Builds a request to get access token.
+
+        Args:
+            login (str): The login.
+            password (str): The password.
+            client (Client): The client.
+            code (Optional[str]): Code from VK/SMS (default value = None).
+            captcha (Optional[Tuple[str, str]]): Captcha (default value = None).
+        """
+        url: str = "https://oauth.vk.com/token"
+
+        params: Dict = {
+            "grant_type": "password",
+            "client_id": client.client_id,
+            "client_secret": client.client_secret,
+            "username": login,
+            "password": password,
+            "scope": "audio,offline",
+            "2fa_supported": 1,
+            "force_sms": 1,
+            "v": DEFAULT_VERSION,
+        }
+        if captcha:
+            params.setdefault("captcha_sid", captcha[0])
+            params.setdefault("captcha_key", captcha[1])
+
+        if code:
+            params.setdefault("code", code)
+
+        request: VkApiRequest = VkApiRequest(
+            method="post",
+            url=url,
+            params=params,
+        )
+
+        request.fill_user_agent(client.user_agent)
+
+        return request
+
+    @staticmethod
+    def build_req_2fa_code(client: Client, sid: Union[str, int]) -> VkApiRequest:
+        """
+        Builds a request to get access token.
+
+        Args:
+            sid (Union[str, int]): Sid from VK.
+        """
+        url: str = "https://api.vk.com/method/auth.validatePhone"
+
+        params: Dict = {
+            "sid": str(sid),
+            "v": DEFAULT_VERSION,
+        }
+
+        request: VkApiRequest = VkApiRequest(
+            method="post",
+            url=url,
+            params=params,
+        )
+
+        request.fill_user_agent(client.user_agent)
+
+        return request
+
+    # Common section
+
+    @staticmethod
+    def build_from_base_request(method: str, url: str, params: dict) -> VkApiRequest:
         """
         Builds a request from a base request.
 
@@ -94,9 +170,7 @@ class VkApiRequestBuilder:
             params["album_id"] = playlist_id
             params["access_key"] = access_key
 
-        return cls.build_from_base_request(
-            method="get", url="audio.get", params=params
-        )
+        return cls.build_from_base_request(method="get", url="audio.get", params=params)
 
     @classmethod
     def build_req_search(

